### PR TITLE
KBV-360  validate client

### PR DESCRIPTION
**Pull request by idrop**
### What changed

* Validate that the issuer and subject for the JWT in the token request match the client id
* Update expected audience and credential issuer URIs in parameter store
* !Ref the KMS signing key directly
* Remove the IssueCredentialFunctionRole and map the policy on the IsseeCredentialFunction directly - tidying up KMS sign and SSM policies

### Why did it change

To fix integration with IPV Core in staging

### Testing
This has been tested on a dev stack, where a complete journey can be completed.

### Issue tracking

- [KBV-360](https://govukverify.atlassian.net/browse/KBV-360)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: infrastructure/lambda/template.yaml [hunk 1]
```diff
@@ -346,66 +338,6 @@ Resources:
               Resource:
                 - !Sub "arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/${AWS::StackName}/clients/*"
 
-  IssueCredentialFunctionRole:
```
Comment (idrop) [inline infrastructure/lambda/template.yaml:349]: This role is now inline in the IssueCredentialFunction, like the other lambdas

File: infrastructure/lambda/template.yaml [hunk 2]
```diff
@@ -415,7 +347,29 @@ Resources:
...
+            TableName: !Ref AddressTable
+        - Statement:
+            Effect: Allow
+            Action:
+              - "kms:Sign"
+            Resource: !Sub "arn:aws:kms:${AWS::Region}:${AWS::AccountId}:key/${AddressCriVcSigningKey}"
+        - SQSSendMessagePolicy:
+            QueueName: !ImportValue AuditEventQueueName
+        - SSMParameterReadPolicy:
+            ParameterName: !Sub "${AWS::StackName}/*"
```
Comment (idrop) [inline infrastructure/lambda/template.yaml:365]: Ooh, what do you think of this? These params aren't secret

File: infrastructure/lambda/template.yaml [hunk 3]
```diff
@@ -460,13 +414,6 @@ Resources:
             Action:
               - 'kms:*'
             Resource: '*'
-          - Sid: "Allow use of the key for digital signing"
```
Comment (idrop) [inline infrastructure/lambda/template.yaml:463]: This is now set on the IssueCredentialFunction

File: infrastructure/lambda/template.yaml [hunk 4]
```diff
@@ -757,7 +697,7 @@ Resources:
     Properties:
       Name: !Sub "/${AWS::StackName}/verifiableCredentialKmsSigningKeyId"
       Type: String
-      Value: !FindInMap [ VerifiableCredentialKmsSigningKeyMapping, Environment, !Ref 'Environment' ]
+      Value: !Ref AddressCriVcSigningKey
```
Comment (idrop) [inline infrastructure/lambda/template.yaml:700]: no need to refer to a mapping, `!Ref AddressCriVcSigningKey` gives the KMS key id

File: infrastructure/lambda/template.yaml [hunk 5]
```diff
@@ -788,6 +728,13 @@ Resources:
       FunctionName: !GetAtt AddressFunction.Arn
       Principal: apigateway.amazonaws.com
 
+  IssueCredentialFunctionPermission:
+    Type: AWS::Lambda::Permission
+    Properties:
+      Action: lambda:InvokeFunction
+      FunctionName: !GetAtt IssueCredentialFunction.Arn
+      Principal: apigateway.amazonaws.com
+
```
Comment (idrop) [inline infrastructure/lambda/template.yaml:737]: just moved the permission to where the other permissions are in the template

File: infrastructure/lambda/template.yaml [hunk 6]
```diff
@@ -797,61 +744,7 @@ Resources:
 
 Outputs:
   AddressApiBaseUrl:
+    Description: "Base url of the Address CRI API"
     Value: !Sub "https://${AddressApi}.execute-api.${AWS::Region}.amazonaws.com/${Environment}/"
     Export:
       Name: !Sub ${AWS::StackName}-AddressApiBaseUrl
-
```
Comment (idrop) [inline infrastructure/lambda/template.yaml:803]: I've dropped all these outputs apart from the base url as they're not necessary - please argue the case to keep some or all though! 😄

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/JWTVerifier.java [hunk 7]
```diff
@@ -31,24 +32,52 @@
...
+                        JWTClaimNames.AUDIENCE,
+                        JWTClaimNames.JWT_ID);
+        JWTClaimsSet expectedClaimValues =
+                new JWTClaimsSet.Builder()
+                        .issuer(clientID.getValue())
+                        .subject(clientID.getValue())
+                        .audience(clientAuthenticationConfig.get("audience"))
+                        .build();
+        verifyJWT(clientAuthenticationConfig, signedJWT, requiredClaims, expectedClaimValues);
+    }
```
Comment (idrop) [inline lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/JWTVerifier.java:68]: this is the main reason for this PR - to indicate the required Claims and expected claim values separately for the Auth Jar JWT verification, and the private_key_jwt verification.

This allows us to assert that client id, iss and sub are the same for the private_key_jwt method.

File: infrastructure/lambda/template.yaml [hunk 8]
```diff
@@ -407,7 +347,29 @@ Resources:
...
+            TableName: !Ref AddressTable
+        - Statement:
+            Effect: Allow
+            Action:
+              - "kms:Sign"
+            Resource: !Sub "arn:aws:kms:${AWS::Region}:${AWS::AccountId}:key/${AddressCriVcSigningKey}"
+        - SQSSendMessagePolicy:
+            QueueName: !ImportValue AuditEventQueueName
+        - SSMParameterReadPolicy:
+            ParameterName: !Sub "${AWS::StackName}/*"
```
Comment (idrop) [inline infrastructure/lambda/template.yaml:365]: What do you think of this? These params aren't secret, and I can't think of a reason that a lambda shouldn't have access to all params. If we adopt this on other lambdas it'll reduce boilerplate yaml

**Review by KaraokeStu — APPROVED** [review]
Much cleaner and more readable - great job!

**Pull request merged**